amacc.c

```diff
@@ -568,112 +546,135 @@ void expr(int lev)
-                                      else { expr(Add); *++e = SHR; }
-                                      ty = INT; break;
         case Add:
-            next(); *++e = PSH; expr(Mul);
-            sz = (ty = t) >= PTR2 ? sizeof(int) :
-                                    ty >= PTR ? tsize[ty - PTR] : 1;
-            if (sz > 1) { *++e = PSH; *++e = IMM; *++e = sz; *++e = MUL; }
-            *++e = ADD;
+            next(); expr(Mul);
+            sz = (ty = t) >= PTR2 ? sizeof(int) :  ty >= PTR ? tsize[ty - PTR] : 1;
+            if (*n == Num) n[1] *= sz; 
+            if (*n == Num && *b == Num) n[1] += b[1]; else { *--n = (int) b; *--n = Add; }
             break;
         case Sub:
-            next(); *++e = PSH; expr(Mul);
-            sz = t >= PTR2 ? sizeof(int) :
-                             t >= PTR ? tsize[t - PTR] : 1;
-            if (t == ty && sz > 1) {
-                switch (sz) {
-                case 4: *++e = SUB; *++e = PSH; *++e = IMM; *++e = 2; *++e = SHR; break;
-                default: *++e = SUB; *++e = PSH; *++e = IMM; *++e = sz; *++e = SUB; break;
+            next(); expr(Mul);
+            sz = t >= PTR2 ? sizeof(int) : t >= PTR ? tsize[t - PTR] : 1;
+            if (sz > 1 && *n == Num){ *--n = sz; *--n = Num; --n; *n = (int) (n + 3); *--n = Mul; }
+            if (*n == Num && *b == Num) n[1] -= b[1]; 
+            else { 
+                *--n = (int) b; *--n = Sub; 
+                if (t == ty && sz > 1) {
+                    switch (sz) {
+                    case 4: *--n = 2; *--n = Num; --n; *n = (int) (n + 3);  *--n = Shr; break;
+                    default: *--n = sz; *--n = Num; --n; *n = (int) (n + 3); *--n = Sub; break;
+                    }
                 }
-                ty = INT;
-            } else if (sz > 1) {
-                *++e = PSH; *++e = IMM; *++e = sz; *++e = MUL; *++e = SUB;
-            } else *++e = SUB;
-            ty = t;
+            }
             break;
         case Mul:
-            next(); *++e = PSH; expr(Inc); *++e = MUL; ty = INT;
+            next(); expr(Inc); 
+            if (*n == Num && *b == Num) n[1] *= b[1]; else { *--n = (int) b; *--n = Mul; } 
+            ty = INT;    
             break;
         case Inc:
         case Dec:
-            if (*e == LC) { *e = PSH; *++e = LC; }
-            else if (*e == LI) { *e = PSH; *++e = LI; }
-            else fatal("bad lvalue in post-increment");
-            sz = ty >= PTR2 ? sizeof(int) :
-                              ty >= PTR ? tsize[ty - PTR] : 1;
-            *++e = PSH; *++e = IMM; *++e = sz;
-            *++e = (tk == Inc) ? ADD : SUB;
-            *++e = (ty == CHAR) ? SC : SI;
-            *++e = PSH; *++e = IMM; *++e = sz;
-            *++e = (tk == Inc) ? SUB : ADD;
+            sz = ty >= PTR2 ? sizeof(int) : ty >= PTR ? tsize[ty - PTR] : 1;
+            if (*n == Load) *n = tk; else { fatal("bad lvalue in post-increment"); }
```

Comment:
[style] Should be `if (*n == Load) *n = tk; else fatal("bad lvalue in post-increment");`.